soweego/linker/linking_strategies.py

```diff
@@ -47,90 +50,134 @@ def baseline(source, target, strategy, output_dir):
+def perfect_name_match(source_dataset, target_entity: BaseEntity) -> dict:
     """Given an iterable of dictionaries ``{string: identifier}``,
     match perfect strings and return a dataset ``{id: id}``.
 
     This strategy applies to any object that can be
     treated as a string: names, links, etc.
     """
+    db_manager = DBManager()
+    session = db_manager.new_session()
     matched = {}
-    merged = defaultdict(list)
-    for dataset in datasets:
-        for string, identifier in dataset.items():
-            merged[string].append(identifier)
-    for string, identifiers in merged.items():
-        if len(identifiers) > 1:
-            LOGGER.debug("'%s': it's a match! %s -> %s",
-                         string, identifiers[0], identifiers[1])
-            matched[identifiers[0]] = identifiers[1]
+
+    for label, qid in source_dataset.items():
+        for res in session.query(target_entity).filter(target_entity.name == label).all():
+            if matched.get(qid):
+                LOGGER.warning(
+                    '%s - %s has already a perfect name match' % (qid, label))
+            matched[qid] = res.catalog_id
+
     return matched
 
 
-def similar_link_match(source, target) -> dict:
-    """Given 2 dictionaries ``{link: identifier}``,
+def similar_link_match(source, target: BaseLinkEntity) -> dict:
+    """Given a dictionaries ``{link: identifier} and a BaseLinkEntity``,
     match similar links and return a dataset ``{source_id: target_id}``.
 
     We treat links as natural language:
     similarity means that a pair of links share a set of keywords.
 
     This strategy only applies to URLs.
     """
-    return perfect_string_match((_process_links(source), _process_links(target)))
+    return similar_name_match(source, target, url_utils.tokenize)
 
 
-def similar_name_match(source, target) -> dict:
-    """Given 2 dictionaries ``{person_name: identifier}``,
+def similar_name_match(source, target, tokenize) -> dict:
+    """Given a dictionaries ``{person_name: identifier}, a BaseEntity and a tokenization function``,
     match similar names and return a dataset ``{source_id: target_id}``.
 
     This strategy only applies to people names.
     """
-    return perfect_string_match((_process_names(source), _process_names(target)))
+    matches = defaultdict(list)
+    to_exclude = set()
+
+    db_manager = DBManager()
+
+    for label, qid in source.items():
+        if not label:
+            continue
+
+        to_exclude.clear()
+
+        tokenized = tokenize(label)
+        if len(tokenized) <= 1:
+            continue
+
+        boolean_search = ' '.join(map('+{0}'.format, tokenized))
+        natural_search = ' '.join(tokenized)
+        session = db_manager.new_session()
+
+        # NOTICE: sets of size 1 are always exluded
+        # Looks for sets equal or bigger containing our tokens
+        ft_search = target.tokens.match(boolean_search)
+        for res in session.query(target).filter(ft_search).all():
```

Comment:
**N.B.:** command for full-text index query.

Reply:
@marfox Can you please tell me more about this?